x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/RestEsqlQueryActionTests.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.action;
+
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.rest.FakeRestRequest;
+import org.elasticsearch.xpack.esql.version.EsqlVersion;
+import org.hamcrest.Matcher;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.action.RestEsqlQueryAction.CLIENT_META;
+import static org.elasticsearch.xpack.esql.action.RestEsqlQueryAction.PRODUCT_ORIGIN;
+import static org.elasticsearch.xpack.esql.action.RestEsqlQueryAction.defaultVersionForOldClients;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+
+public class RestEsqlQueryActionTests extends ESTestCase {
+    public void testNoVersionForNoClient() {
+        assertEsqlVersion(null, null, nullValue(String.class));
+    }
+
+    public void testNoVersionForAlreadySet() {
+        EsqlQueryRequest esqlRequest = new EsqlQueryRequest();
+        esqlRequest.esqlVersion("whatever");
+        FakeRestRequest restRequest = new FakeRestRequest();
+        restRequest.getHttpRequest().getHeaders().put(CLIENT_META, List.of("es=8.13.0"));
+        defaultVersionForOldClients(esqlRequest, restRequest);
+        assertThat(esqlRequest.esqlVersion(), equalTo("whatever"));
+    }
+
+    public void testNoVersionForNewClient() {
+        assertEsqlVersion("es=8.14", randomProduct(), nullValue(String.class));
```

Comment:
super nit: could use future versions as well, 8.15, ..., 8.21 (an 8.2x might be nice since we do some substring logic)

Reply:
++ - I'll randomize to 8.2x as well.